자동차경주게임: 1단계

iamkyu opened the pull request:
- 초간단 자동차 경주 게임을 구현한다.
- 주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
- 사용자는 몇 대의 자동차로 몇 번의 이동을 할 것인지를 입력할 수 있어야 한다.
- 전진하는 조건은 0에서 9 사이에서 random 값을 구한 후 random 값이 4이상일 경우이다.
- 자동차의 상태를 화면에 출력한다. 어느 시점에 출력할 것인지에 대한 제약은 없다.

iamkyu committed:
자동차경주게임: 1단계

javajigi commented on src/main/java/race/Drawer.java:
```diff
@@ -0,0 +1,27 @@
+}
+
+class LineDrawer implements Drawer {
+    private String format = "-";
```
상수 값 convention을 따라라.
상수 값은 static final로 구현하면 변수 이름을 모두 대문자로 구현하는 것이 관례이다.

javajigi commented on src/main/java/race/Main.java:
```diff
@@ -0,0 +1,32 @@
+        int cars = scanner.nextInt();
+
+        System.out.println("시도할 회수는 몇 회 인가요?");
+        int iteration = scanner.nextInt();
```
View 관련 로직도 별도의 클래스로 분리하면 좋을 것 같아요.

javajigi commented on src/main/java/race/MoveRule.java:
```diff
@@ -0,0 +1,12 @@
+class DefaultMoveRule implements MoveRule {
+    @Override
+    public boolean canMove(int randomValue) {
+        return randomValue >= 4;
```
값을 하드코딩하지 마라.
상수(static final)를 만들고 이름을 부여해 이 변수의 역할이 무엇인지 의도를 드러내라.

javajigi commented on src/main/java/race/MoveRule.java:
```diff
@@ -0,0 +1,12 @@
+package race;
+
+interface MoveRule {
+    boolean canMove(int randomValue);
```
인터페이스 추가 👍

javajigi commented on src/main/java/race/Car.java:
```diff
@@ -0,0 +1,22 @@
+        initDistance();
+    }
+
+    void move() {
```
메소드 인자로 MoveRule을 인자로 받아 구현하는 것은 어떤가요?

javajigi commented on src/main/java/race/RacingGame.java:
```diff
@@ -0,0 +1,52 @@
+    private MoveRule rule;
+    private Drawer drawer;
+    
+    private Random random;
```
인스턴스 변수가 너무 많다. 인스턴스 변수를 줄여본다.
객체지향 설계 원칙의 "규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다." 원칙을 따라본다. 즉, 2개까지 허용한다.

javajigi commented on src/test/java/race/DrawerTest.java:
```diff
@@ -0,0 +1,41 @@
+        //then
+        Assert.assertThat(outputCapture.toString(), is("---\n\n"));
+    }
+}
```
View에 대한 테스트는 어려워 요구하지 않는데 💯

javajigi reviewed (changes requested):
구현 잘 했네요. 객체간의 관계를 RacingGame에 집중했는데요. 분산할 방법을 찾아보면 좋을 것 같아요.

iamkyu committed:
[코드리뷰반영] 자동차경주게임: 1단계

- 매직넘버를 상수로 변경
- 일부 상수 값 Convention에 따라 변수명 변경
- RaceGame에 집중되어 있는 객체 관계를 나누는 과정에서 일부 클래스 추가됨 (GameResult, Recorder)
- GameView 클래스 추가
- 리팩토링 과정에서 일부 용어를 변경
    - Car -> RaceCar
    - iteration -> labs
    - MoveRule -> RaceRule

iamkyu commented:
@javajigi 
리뷰 감사합니다. 다음과 같은 내용을 반영했습니다.

- 매직넘버를 상수로 변경
- 일부 상수 값 Convention에 따라 변수명 변경
- RaceGame에 집중되어 있는 객체 관계를 나누는 과정에서 일부 클래스 추가됨 (GameResult, Recorder)
- GameView 클래스 추가
- 리팩토링 과정에서 일부 용어를 변경 
    - Car -> RaceCar
    - iteration -> labs
    - MoveRule -> RaceRule
